kaggle_environments/envs/werewolf/eval/metrics.py

```diff
@@ -0,0 +1,875 @@
+            for g in self.games:
+                yield LightGame(
+                    [LightPlayer(p.agent.display_name, p.role.team) for p in g.players],
+                    g.winner_team
+                )
+        
+        samples_iterator = self._generate_bootstrap_samples(light_games_iter(), num_samples)
+
+        worker = functools.partial(_openskill_bootstrap_worker, model=self.openskill_model)
+
+        with ProcessPoolExecutor(max_workers=os.cpu_count()) as executor:
+            results = list(tqdm(executor.map(worker, samples_iterator), total=num_samples, desc="OpenSkill Bootstrap"))
+
+        bootstrapped_mus = defaultdict(list)
+        all_agents = list(self.metrics.keys())
+
+        default_mu = self.openskill_model.rating().mu
+
+        for sample_ratings in results:
+            for agent in all_agents:
+                if agent in sample_ratings:
+                    bootstrapped_mus[agent].append(sample_ratings[agent])
+                else:
+                    bootstrapped_mus[agent].append(default_mu)
+
+        for agent, values in bootstrapped_mus.items():
+            if len(values) > 1:
+                self.metrics[agent].openskill_mu_std = float(np.std(values, ddof=1))
+
+    def evaluate(self, gte_samples=3, elo_samples=3, openskill_samples=3):
+        for game in self.games:
+            for player in game.players:
+                agent_name = player.agent.display_name
+                if self.metrics[agent_name].agent_name is None:
+                    self.metrics[agent_name].set_agent_name(agent_name)
+                won = 1 if player.role.team == game.winner_team else 0
+                survived = 1 if player.alive else 0
+                self.metrics[agent_name].wins.append(won)
+                self.metrics[agent_name].wins_by_role[player.role.name].append(won)
+                self.metrics[agent_name].survival_scores.append(survived)
+                self.metrics[agent_name].survival_by_role[player.role.name].append(survived)
+            irp_results, vss_results = game.iterate_voting_mini_game()
+            for agent_name, score in irp_results:
+                self.metrics[agent_name].irp_scores.append(score)
+            for agent_name, score in vss_results:
+                self.metrics[agent_name].vss_scores.append(score)
+
+        # Elo
+        final_elos = self._compute_elo_ratings(self.games)
+        for agent, rating in final_elos.items():
+            self.metrics[agent].elo = rating
+
+        # OpenSkill
+        if OPENSKILL_AVAILABLE and self.openskill_model:
+            final_openskill_ratings = self._compute_openskill_ratings(self.games, self.openskill_model)
+            for agent, rating in final_openskill_ratings.items():
+                self.metrics[agent].openskill_rating = rating
+
+        self._bootstrap_elo(num_samples=elo_samples)
+        self._bootstrap_openskill(num_samples=openskill_samples)
+        self._run_gte_evaluation(num_samples=gte_samples)
```

Comment:
Log processing is done twice through iterate_voting_mini_game here and in line 463, this can be consolidated to reduce redundency?